test/fw/ptl/lib/ptl_comm.py

```diff
@@ -0,0 +1,259 @@
+# organizations to create proprietary derivative works of OpenPBS and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair's trademarks, including but not limited to "PBS™",
+# "OpenPBS®", "PBS Professional®", and "PBS Pro™" and Altair's logos is
+# subject to Altair's trademark licensing policies.
+
+
+import copy
+import datetime
+import logging
+import os
+import re
+import socket
+import string
+import sys
+import time
+try:
+    from ptl.lib.pbs_ifl import *
+    API_OK = True
+except:
+    try:
+        from ptl.lib.pbs_ifl_mock import *
+    except:
+        sys.stderr.write("failed to import pbs_ifl, run pbs_swigify " +
+                         "to make it\n")
+        raise ImportError
+    API_OK = False
+
+from ptl.lib.pbs_testlib import *
+
+
+class Comm(PBSService):
+
+    """
+    PBS ``Comm`` configuration and control
+    """
+
+    """
+    :param name: The hostname of the Comm. Defaults to current hostname.
+    :type name: str
+    :param attrs: Dictionary of attributes to set, these will override
+                  defaults.
+    :type attrs: dictionary
+    :param pbsconf_file: path to config file to parse for PBS_HOME,
+                         PBS_EXEC, etc
+    :type pbsconf_file: str or None
+    :param snapmap: A dictionary of PBS objects (node,server,etc) to
+                    mapped files from PBS snap directory
+    :type snapmap: dictionary
+    :param snap: path to PBS snap directory (This will override snapmap)
+    :type snap: str or None
+    :param server: A PBS server instance to which this Comm is associated
+    :type server: str
+    :param db_access: set to either file containing credentials to DB access or
+                      dictionary containing {'dbname':...,'user':...,
+                      'port':...}
+    :type db_access: str or dictionary
+        """
+    dflt_attributes = {}
+
+    def __init__(self, server, name=None, attrs={}, pbsconf_file=None,
+                 snapmap={}, snap=None, db_access=None):
+        self.server = server
+        if snap is None and self.server.snap is not None:
+            snap = self.server.snap
+        if (len(snapmap) == 0) and (len(self.server.snapmap) != 0):
+            snapmap = self.server.snapmap
+        PBSService.__init__(self, name, attrs, self.dflt_attributes,
+                            pbsconf_file, snapmap, snap)
+        _m = ['Comm ', self.shortname]
+        if pbsconf_file is not None:
+            _m += ['@', pbsconf_file]
+        _m += [': ']
+        self.logprefix = "".join(_m)
+        self.conf_to_cmd_map = {
+            'PBS_COMM_ROUTERS': '-r',
+            'PBS_COMM_THREADS': '-t'
+        }
+        self.pi = PBSInitServices(hostname=self.hostname,
+                                  conf=self.pbs_conf_file)
+
+    def __del__(self):
+        del self.__dict__
+
+    def isUp(self):
```

Comment:
@vishwaks All daemons use the common multiple service of PBSService class through their own service method.

Reply:
@vishwaks I think its there because we don't have `isUp` in PBSService class but we have only `_isUp` so each daemon class defines `isUp` by itself.

@birbalsain I think what vishwa is saying (and I agree with him) we should define common method in PBSService and override these methods only if we need it. And AFAIK only server class needs to override (this specific isUp()) as server has to call pbs_disconnect() or run qterm instead running kill command.